ipa-core/src/protocol/boolean/and.rs

```diff
@@ -0,0 +1,42 @@
+use std::iter::zip;
+
+use crate::{
+    error::Error,
+    ff::boolean::Boolean,
+    protocol::{basics::SecureMul, context::Context, step::BitOpStep, RecordId},
+    secret_sharing::{replicated::semi_honest::AdditiveShare, BitDecomposed, FieldSimd},
+};
+
+/// Matrix bitwise AND for use with vectors of bit-decomposed values
+///
+/// ## Errors
+/// Propagates errors from the multiplication protocol.
+/// ## Panics
+/// Panics if the bit-decomposed arguments do not have the same length.
+//
+// Supplying an iterator saves constructing a complete copy of the argument
+// in memory when it is a uniform constant.
+pub async fn bool_and<'a, C, BI, const N: usize>(
+    ctx: C,
+    record_id: RecordId,
+    a: &BitDecomposed<AdditiveShare<Boolean, N>>,
+    b: BI,
+) -> Result<BitDecomposed<AdditiveShare<Boolean, N>>, Error>
+where
+    C: Context,
+    BI: IntoIterator,
+    <BI as IntoIterator>::IntoIter: ExactSizeIterator<Item = &'a AdditiveShare<Boolean, N>> + Send,
+    Boolean: FieldSimd<N>,
+    AdditiveShare<Boolean, N>: SecureMul<C>,
+{
+    let b = b.into_iter();
+    assert_eq!(a.len(), b.len());
+
+    BitDecomposed::try_from(
+        ctx.parallel_join(zip(a.iter(), b).enumerate().map(|(i, (a, b))| {
+            let ctx = ctx.narrow(&BitOpStep::Bit(i));
```

Comment:
We might need a new name for this step. This isn't really operating on "bits" anymore, this is now operating on vectorized objects.

Reply:
My hope was that in many cases, vectorized protocols could read like unvectorized protocols -- the code says what to do for each record, and under the hood that magically gets repeated for a vector of records.

Do you have a suggestion how to name it that would be clearer, without making things too verbose?

Reply:
I see what you're saying. Yeah, I guess it is operating on a "bit" in some sense. Yeah, let's keep this name.